### ISSUE TITLE
Angular 4 fails to provide MockBackend

### ISSUE THREAD

**Issue by halamah** [angular/angular#15521]
**I'm submitting a ...**  (check one with "x")
```
[x] bug report => search github for a similar issue or PR before submitting
[ ] feature request
[ ] support request => Please do not submit support request here, instead see https://github.com/angular/angular/blob/master/CONTRIBUTING.md#question
```

**Current behavior**

Exception if adding MockBackend to a list of AppModule providers: 

**"Uncaught Error: Encountered undefined provider! Usually this means you have a circular dependencies (might be caused by using 'barrel' index.ts files."**

This behavior since Angular 4.0 RC4/5. Mocking this way in 4.0 RC3 and earlier working just fine.

Exception in CompileMetadataResolver.prototype._getProvidersMetadata function
...
else if (provider === void 0) {
    _this._reportError(syntaxError("Encountered undefined provider! Usually this means you have a circular dependencies (might be caused by using 'barrel' index.ts files."));
}
...

**Expected behavior**

FakeProvider is instantiated and MockBackend instance successfully injected.

**Minimal reproduction of the problem with instructions**

Add MockBackend to a list of AppModule providers.
...
import {MockBackend} from "@angular/http/testing";
...
providers: [
        ...
       MockBackend, // <--
        //FakeProvider,
        //BaseRequestOptions
],
...

**What is the motivation / use case for changing the behavior?**

Mock an Http service for testing purpose and backend-less development

**Please tell us about your environment:**

* **Angular version:** 4.0.0

* **Browser:** [all | Chrome XX | Firefox XX | IE XX | Safari XX | Mobile Chrome XX | Android X.X Web Browser | iOS XX Safari | iOS XX UIWebView | iOS XX WKWebView ]

 
* **Language:** Typescript 2.2

* **Node (for AoT issues):** `node --version` =  6.6.0

*     webpack: "^2.3.2", webpack-dev-server: "^2.4.2", webpack-merge: "^4.1.0"

**Comment by DzmitryShylovich** [comment]
Please add a small reproduction

**Comment by halamah** [comment]
Done, added.

**Comment by DzmitryShylovich** [comment]
http://plnkr.co/edit/MAVE5Mq3aaFFbUauXmRh?p=preview

**Comment by halamah** [comment]
Although that happened after updating to Angular 4.0 RC4/5

provider = __webpack_require__.i(__WEBPACK_IMPORTED_MODULE_0__angular_core__["_46" /* resolveForwardRef */])(provider);
...
else if (provider === void 0) {
_this._reportError(syntaxError("Encountered undefined provider! Usually this means you have a circular dependencies (might be caused by using 'barrel' index.ts files."));
}

Could it be a webpack related issue?

**Comment by halamah** [comment]
Set all angular components to latest version 4.0, but reverted http module to 4.0 RC3 - and MockBackend provider now works as expected.

Are there any changes I need to consider when moving to angular 4.0?

**Comment by DzmitryShylovich** [comment]
There has been no changes.
I've just tested it using [this starter](https://github.com/qdouble/angular-webpack2-starter) and haven't got any errors. Here's my code
```
import {Component, NgModule} from '@angular/core'
import {MockBackend} from '@angular/http/testing'
import {BrowserModule} from '@angular/platform-browser'
import {ConnectionBackend} from '@angular/http';

@Component({
  selector: 'my-app',
  template: `
    <div>
      <h2>Hello {{name}}</h2>
    </div>
  `,
})
export class App {
  name:string;
  constructor(backend: ConnectionBackend) {
    console.log(backend);
    this.name = 'Angular!'
  }
}

@NgModule({
  imports: [ BrowserModule ],
  declarations: [ App ],
  bootstrap: [ App ],
  providers: [{provide: ConnectionBackend, useClass: MockBackend}]
})
export class AppModule {}
```

**Comment by rerickson1** [comment]
I believe I am seeing a similar issue.  
If I build with AOT (ng build --prod):
ERROR in ./src/$$_gendir/app/app.module.ngfactory.ts
Module not found: Error: Can't resolve '@angular/http/testing/src/mock_backend' in 'C:\git\web_lcm_gatwick\src\$$_gendir\app'
 @ ./src/$$_gendir/app/app.module.ngfactory.ts 60:0-67
 @ ./src/main.ts
 @ multi ./src/main.ts

If I comment out MockBackend in my providers the issue goes away.  If I disable AOT, the issue goes away.

Simple reproduction with angular-cli 1.0.0:

1. ng new ngcli
2. edit app.module.ts

```import { BrowserModule } from '@angular/platform-browser';
import { NgModule } from '@angular/core';
import { FormsModule } from '@angular/forms';
import { HttpModule } from '@angular/http';
import { MockBackend } from "@angular/http/testing";

import { AppComponent } from './app.component';

@NgModule({
  declarations: [
    AppComponent
  ],
  imports: [
    BrowserModule,
    FormsModule,
    HttpModule
  ],
  providers: [
      MockBackend
  ],
  bootstrap: [AppComponent]
})
export class AppModule { }
```

3.  ```ng build --prod``` produces error:

ERROR in ./src/$$_gendir/app/app.module.ngfactory.ts
Module not found: Error: Can't resolve '@angular/http/testing/src/mock_backend' in 'C:\git\ngcli\src\$$_gendir\app'
 @ ./src/$$_gendir/app/app.module.ngfactory.ts 23:0-66
 @ ./src/main.ts
 @ multi ./src/main.ts

4.  if you run ```ng build``` everything works fine

**Comment by vascofernandes** [comment]
same problem, MockBackend not available in aot.

**Comment by beelaal** [comment]
How i can run Angular 4 with MVC Asp.net ?? anybody please help

**Comment by rerickson1** [comment]
Just tried this again with Angular version 4.0.3 and the issue still exists.  
Using my reproduction steps above >https://github.com/angular/angular/issues/15521#issuecomment-290118657 

 `ng build --prod` or `ng serve --prod` produces the error.

**Comment by chi-lee** [comment]
Same problem using Angular CLI 1.0.0 and Angular 4.0.0

**Comment by mertyildiran** [comment]
I'm getting this error:

```
ERROR in ./src/$$_gendir/app/app.module.ngfactory.ts
Module not found: Error: Can't resolve '@angular/http/testing/src/mock_backend' in '/home/mertyildiran/Documents/Tower/src/$$_gendir/app'
 @ ./src/$$_gendir/app/app.module.ngfactory.ts 26:0-66
 @ ./src/main.ts
 @ multi ./src/main.ts
```

You can reproduce the error with using https://github.com/mertyildiran/Tower/tree/4c6c48a556b073d9c17df703e07cc85c4d28d0cd

**Comment by FreezeSoul** [comment]
I‘ve also got the same error on ionic2/3  with " --prod"!

> Uncaught Error: Cannot find module "@angular/http/testing/src/mock_backend"

**Comment by pain0486** [comment]
So what version will this be released in?

**Comment by nicholasserra** [comment]
Confirmed fix in 4.1.2. Thanks everyone!

**Comment by pearldiver** [comment]
The issue is closed but i still get the error, i just upgraded to angular 4 (4.1.3) 2 days before, all the production code works fine, but all the tests are down, on reason is the MockBackend imported form '@angular/http/testing'  is undefined.
My project don't use cli, the test is started by Karma and webpack.
the curious thing is all other angular module import works except for testing include '@angular/http/testing' and '@angular/core/testing' 
Any ideas? thanks!

**Comment by sittingbool** [comment]
Same here, also webpack

**Comment by iturn** [comment]
+1 confirmed ng4.0.0, building --prod

**Comment by madhavsharmaagra** [comment]
I have started to use Angular 5 but it still seems that the deep imports are not working yet. My IDE automatically predicts and add import to the related but to the deep level. By removing the deep link, the app started to work fine. For example,

import { ElementRef } from '@angular/core'; // works fine but on the other hand, 

import { ElementRef } from '@angular/core/src/linker/element_ref'; // throws error.

I made it to work just by removing the rest of the deep link and leaving only '@angular/core';
